src/right-field/results-table/ResultsTable.jsx

```diff
@@ -0,0 +1,129 @@
+}
+
+const QueryEngine = require('@comunica/query-sparql').QueryEngine;
+const myEngine = new QueryEngine()
+
+/**
+ * 
+ * @param {query} props.query The query (as defined in the config file) that should be executed and results displayed in the table. 
+ * @returns {Component} A React component giving a structural representation of the query results.  
+ */
+function ResultsTable(props){
+    const selectedquery = props.selectedquery
+    const [results, setResults] = useState([])
+    const [variables, setVariables] = useState([])
+    let adder = (item) => setResults((old) => {return [...old, item]})
+    let onqueryChanged = () => {
+      if(selectedquery){
+        setResults([])
+        executequery(selectedquery, adder, setVariables)
+      }
+    }
+
+    if(props.refreshButton.current){
+      props.refreshButton.current.onclick = onqueryChanged
+    }
+
+    useEffect(() => {
+        onqueryChanged()
+    }, [selectedquery])
+
+    return(
+        <div className="results-table">
+            {!selectedquery && <label>Please select a query.</label>}
+            {selectedquery && 
+            <Grid style={{td: {"text-align": "center"}, th: {"text-align": "center"}}} 
+            data={results} 
+            fixedHeader={true}
+            height={"100%"}
+            autoWidth="true" columns={variables}/>
+            }
+        </div>
+    )
+}
+
+/**
+ * A function that executes a given query and processes every result as a stream based on the functions provided. 
+ * @param {query} query the query which gets executed 
+ * @param {Function} adder a function which handles what happens with every variable result  
+ * @param {Function} variableSetter a function which handles what happens with every variable name 
+*/
+function executequery(query, adder, variableSetter){
+    return fetch(`${config.queryFolder}${query.queryLocation}`).then(result => {
+      result.text().then(q => {
+        const execution = myEngine.queryBindings(
+          q,
+          {sources:query.sources}
+        )
+        handlequeryExecution(execution, adder, variableSetter)
+      })
+    }).catch(handlequeryFetchFail)
+  }
+
+
+/**
+ * A function that given a BindingStream processes every result as a stream based on the functions provided. 
+ * @param {Promise<BindingStream>} execution   
+ * @param {Function} adder a function which handles what happens with every variable result  
+ * @param {Function} variableSetter a function which handles what happens with every variable name 
+ */
+function handlequeryExecution(execution, adder, variableSetter){
+  execution.then(bindingStream => {
+    bindingStream.on('data', (binding) => {
+      let triple = []
+      let variables = []
+      let keys = binding.keys()
+      let key = keys.next()
+      while(!key.done){
+          triple.push(binding.get(key.value.value).id)
+          variables.push(key.value.value)
+          key = keys.next()
+      }   
+      variableSetter(variables)
```

Comment:
Why does this happen for every row "row" in the results? Is not enough to this only once, because the variables remain the same for every "row"?

Reply:
I have indeed forgot to change this. 

I have changed one of the test queries, as an example to this issue. When using Optional variables, when the optional variable is not present, the communica engine doesn't return the variable as a key. 

Currently this is solved as in (this commit)[https://github.com/SolidLabResearch/generic-data-viewer/pull/1/commits/07e5723b1ea097179f2f921e9ea5ed6cf34b9351].

Reply:
For extra clarity

![image](https://github.com/SolidLabResearch/generic-data-viewer/assets/108480125/a32bbf8a-2343-44b2-93b6-06aac793fff0)
This is the result using the current commit. 

![image](https://github.com/SolidLabResearch/generic-data-viewer/assets/108480125/a8f0fe0d-b007-4f45-ab9b-7415a49833ea)

this would be the result using once.